### PR TITLE
Turbopack: import webpack loader rules conditions

### PR DESCRIPTION
**Pull request by sokra**
### What?

* remove custom next-* conditions
* add `foreign` condition
* allow `false` in turbo.rules
* improve schema for turbo.rules

### How?

Closes PACK-2913

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| buildDuration | 14.9s | 15s | ⚠️ +107ms |
| buildDurationCached | 8.2s | 6.8s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +2.32 kB |
| nextStartRea..uration (ms) | 413ms | 411ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 241 B | N/A |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99 kB | 99 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 522 B | N/A |
| Overall change | 541 B | 541 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.4 kB | 95.4 kB | N/A |
| page.js gzip | 3.06 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 97 kB | 97 kB | ✓ |
| app-page-tur..prod.js gzip | 98.8 kB | 98.8 kB | ✓ |
| app-page-tur..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-page.run...dev.js gzip | 144 kB | 144 kB | ✓ |
| app-page.run..prod.js gzip | 91.5 kB | 91.5 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.1 kB | 21.1 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.4 kB | 22.4 kB | ✓ |
| server.runti..prod.js gzip | 51.1 kB | 51.1 kB | ✓ |
| Overall change | 945 kB | 945 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/improve-webpack-rules | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | N/A |
| index.pack gzip | 107 kB | 106 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: accc211df840ca21f0dce8d63d3962a967c99965

File: packages/next-swc/crates/next-core/src/next_edge/context.rs [hunk 1]
```diff
@@ -100,11 +100,13 @@ pub async fn get_edge_resolve_options_context(
...
-    let mut custom_conditions = vec![
-        mode.await?.condition().to_string(),
-        "edge-light".to_string(),
-        "worker".to_string(),
-    ];
+    let mut custom_conditions = vec![mode.await?.condition().to_string()];
+    custom_conditions.extend(
+        NextRuntime::Edge
+            .conditions()
+            .into_iter()
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_edge/context.rs:107]:
```suggestion
            .iter()
```

File: packages/next-swc/crates/next-core/src/next_server/context.rs [hunk 2]
```diff
@@ -157,8 +143,13 @@ pub async fn get_server_resolve_options_context(
     );
     let ty = ty.into_value();
 
-    let mut custom_conditions = vec![mode.await?.condition().to_string(), "node".to_string()];
-    custom_conditions.extend(ty.conditions().iter().map(ToString::to_string));
+    let mut custom_conditions = vec![mode.await?.condition().to_string()];
+    custom_conditions.extend(
+        NextRuntime::NodeJs
+            .conditions()
+            .into_iter()
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_server/context.rs:150]:
```suggestion
            .iter()
```

File: packages/next-swc/crates/next-core/src/next_server/context.rs [hunk 3]
```diff
@@ -354,37 +345,33 @@ pub async fn get_server_module_options_context(
     let enable_postcss_transform = Some(postcss_transform_options.cell());
     let enable_foreign_postcss_transform = Some(postcss_foreign_transform_options.cell());
 
+    let mut conditions = vec![mode.await?.condition().to_string()];
+    conditions.extend(
+        next_runtime
+            .conditions()
+            .into_iter()
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_server/context.rs:352]:
```suggestion
            .iter()
```

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 11cf3e4df7a832524ec2467f1ea5c3a10549ea28

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/parallel-routes-and-interception/parallel-routes-and-interception.test.ts` (turbopack)
- parallel-routes-and-interception > route intercepting > should render modal when paired with parallel routes

<details>
<summary>Expand output</summary>

  ● parallel-routes-and-interception › route intercepting › should render modal when paired with parallel routes

    thrown: "Exceeded timeout of 60000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      774 |       })
      775 |
    > 776 |       it('should render modal when paired with parallel routes', async () => {
          |       ^
      777 |         const browser = await next.browser(
      778 |           '/intercepting-parallel-modal/vercel'
      779 |         )

      at it (e2e/app-dir/parallel-routes-and-interception/parallel-routes-and-interception.test.ts:776:7)
      at describe (e2e/app-dir/parallel-routes-and-interception/parallel-routes-and-interception.test.ts:623:5)
      at fn (lib/e2e-utils.ts:328:5)
      at describe (lib/e2e-utils.ts:321:3)
      at Object.<anonymous> (e2e/app-dir/parallel-routes-and-interception/parallel-routes-and-interception.test.ts:5:19)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by timneutkens — APPROVED** [review]

**Comment by timneutkens** [comment]
* **#64205** <a href="https://app.graphite.dev/github/pr/vercel/next.js/64205?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sokra and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Pull request merged**